x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/ChangePointStates.java

```diff
@@ -0,0 +1,278 @@
+            count++;
+            timestamps = bigArrays.grow(timestamps, count);
+            timestamps.set(count - 1, timestamp);
+            values = bigArrays.grow(values, count);
+            values.set(count - 1, value);
+        }
+
+        void add(LongBlock timestamps, DoubleBlock values) {
+            add(timestamps, values, 0);
+        }
+
+        void add(LongBlock timestamps, DoubleBlock values, int otherPosition) {
+            final int valueCount = timestamps.getValueCount(otherPosition);
+            final int firstIndex = timestamps.getFirstValueIndex(otherPosition);
+            for (int i = 0; i < valueCount; i++) {
+                add(timestamps.getLong(firstIndex + i), values.getDouble(firstIndex + i));
+            }
+        }
+
+        void toIntermediate(Block[] blocks, int offset, DriverContext driverContext) {
+            blocks[offset] = buildTimestampsBlock(driverContext.blockFactory());
+            blocks[offset + 1] = buildValuesBlock(driverContext.blockFactory());
+        }
+
+        // TODO: this needs to output multiple columns or a composite object, not a JSON blob.
+        Block toBlock(BlockFactory blockFactory) {
+            return blockFactory.newConstantBytesRefBlockWith(getChangePoint(), 1);
+        }
+
+        private Block buildTimestampsBlock(BlockFactory blockFactory) {
+            if (timestamps.size() == 0) {
+                return blockFactory.newConstantNullBlock(1);
+            }
+            if (values.size() == 1) {
+                return blockFactory.newConstantLongBlockWith(timestamps.get(0), 1);
+            }
+            try (LongBlock.Builder builder = blockFactory.newLongBlockBuilder((int) timestamps.size())) {
+                builder.beginPositionEntry();
+                for (int id = 0; id < count; id++) {
+                    builder.appendLong(timestamps.get(id));
+                }
+                builder.endPositionEntry();
+                return builder.build();
+            }
+        }
+
+        private Block buildValuesBlock(BlockFactory blockFactory) {
+            if (values.size() == 0) {
+                return blockFactory.newConstantNullBlock(1);
+            }
+            if (values.size() == 1) {
+                return blockFactory.newConstantDoubleBlockWith(values.get(0), 1);
+            }
+            try (DoubleBlock.Builder builder = blockFactory.newDoubleBlockBuilder((int) values.size())) {
+                builder.beginPositionEntry();
+                for (int id = 0; id < count; id++) {
+                    builder.appendDouble(values.get(id));
+                }
+                builder.endPositionEntry();
+                return builder.build();
+            }
+        }
+
+        // TODO: this needs to output multiple columns or a composite object, not a JSON blob.
+        private BytesRef getChangePoint() {
+            // TODO: probably reuse ES|QL sort/orderBy to get results in order
```

Comment:
This is *partly* what I mean by "make this a time series". ESQL would stick a sort before the aggregation operation. Or, well, iterate the documents in time series order. Which it totally can do for time series indices.